File: 3351-maximize-happiness-of-selected-children/solution.go
import (
    "slices"
)

func maximumHappinessSum(happiness []int, k int) int64 {
    slices.SortFunc(happiness, func(a, b int) int {
        return b - a
    })

    var total int64 = 0
    
    for i := 0; i < k; i++ {
        val := happiness[i] - i
        
        if val <= 0 {
            break
        }
        
        total += int64(val)
    }

    return total
}

File: 3351-maximize-happiness-of-selected-children/solution.rs
impl Solution {
    pub fn maximum_happiness_sum(mut happiness: Vec<i32>, k: i32) -> i64 {

        happiness.sort_unstable_by(|a, b| b.cmp(a));
        
        let mut total: i64 = 0;
        let k = k as usize;

        for i in 0..k {

            let val = happiness[i] as i64 - i as i64;

            if val <= 0 {
                break;
            }
            
            total += val;
        }
        
        total
    }
}
